freqtrade/freqai/utils.py

```diff
@@ -0,0 +1,138 @@
+import logging
+from datetime import datetime, timezone
+
+from freqtrade.configuration import TimeRange
+from freqtrade.data.dataprovider import DataProvider
+from freqtrade.data.history.history_utils import refresh_backtest_ohlcv_data
+from freqtrade.exceptions import OperationalException
+from freqtrade.exchange import timeframe_to_seconds
+from freqtrade.exchange.exchange import market_is_active
+from freqtrade.plugins.pairlist.pairlist_helpers import dynamic_expand_pairlist
+
+
+logger = logging.getLogger(__name__)
+
+
+def download_all_data_for_training(dp: DataProvider, config: dict) -> None:
+    """
+    Called only once upon start of bot to download the necessary data for
+    populating indicators and training the model.
+    :param timerange: TimeRange = The full data timerange for populating the indicators
+                                    and training the model.
+    :param dp: DataProvider instance attached to the strategy
+    """
+
+    if dp._exchange is not None:
+        markets = [p for p, m in dp._exchange.markets.items() if market_is_active(m)
+                   or config.get('include_inactive')]
+    else:
+        # This should not occur:
+        raise OperationalException('No exchange object found.')
+
+    all_pairs = dynamic_expand_pairlist(config, markets)
+
+    timerange = get_required_data_timerange(config)
+
+    new_pairs_days = int((timerange.stopts - timerange.startts) / 86400)
+    if not dp._exchange:
+        # Not realistic - this is only called in live mode.
+        raise OperationalException("Dataprovider did not have an exchange attached.")
+    refresh_backtest_ohlcv_data(
+        dp._exchange,
+        pairs=all_pairs,
+        timeframes=config["freqai"]["feature_parameters"].get("include_timeframes"),
+        datadir=config["datadir"],
+        timerange=timerange,
+        new_pairs_days=new_pairs_days,
+        erase=False,
+        data_format=config.get("dataformat_ohlcv", "json"),
+        trading_mode=config.get("trading_mode", "spot"),
+        prepend=config.get("prepend_data", False),
+    )
+
+
+def get_required_data_timerange(
+    config: dict
+) -> TimeRange:
+    """
+    Used to compute the required data download time range
+    for auto data-download in FreqAI
+    """
+    time = datetime.now(tz=timezone.utc).timestamp()
+    data_load_timerange = TimeRange()
+
+    timeframes = config["freqai"]["feature_parameters"].get("include_timeframes")
+
+    max_tf_seconds = 0
+    for tf in timeframes:
+        secs = timeframe_to_seconds(tf)
+        if secs > max_tf_seconds:
+            max_tf_seconds = secs
```

Comment:
if you wanna be fancy - we could also do this via the following comprehension

```suggestion
    max_tf_seconds = max(timeframe_to_seconds(tf) for tf in timeframes)
```

Not sure if that really improves readability though ...